sys/linux/socket.txt

```diff
@@ -317,6 +321,41 @@ ifr_ifru [
 
 define IFREQ_SIZE	sizeof(struct ifreq)
 
+# https://elixir.bootlin.com/linux/v5.3.11/source/include/uapi/linux/if_vlan.h
+vlan_args {
+	cmd		flags[vlan_cmd, intptr]
+	device1		array[int8, 24]
+	u		vlan_u
+	vlan_qos	int8
+}
+
+vlan_u [
+	device2		array[int8, 24]
+	VID		intptr
+	skb_priority	intptr
+	name_type	intptr
+	bind_type	intptr
+	flag		intptr
+]
+
+vlan_cmd = ADD_VLAN_CMD, DEL_VLAN_CMD, SET_VLAN_INGRESS_PRIORITY_CMD, SET_VLAN_EGRESS_PRIORITY_CMD, GET_VLAN_INGRESS_PRIORITY_CMD, GET_VLAN_EGRESS_PRIORITY_CMD, SET_VLAN_NAME_TYPE_CMD, SET_VLAN_FLAG_CMD, GET_VLAN_REALDEV_NAME_CMD, GET_VLAN_VID_CMD
```

Comment:
A better way to represent this (for fuzzer) would be to transpose the structure.
Such unions are frequently used in interfaces, but that's simply due to C's non-expressiveness and amount of duplication that would be required to express this properly in C. What we have here is this enum and parallel values in the union. And I guess there is actually one-to-one correspondence between the vlan_cmd and the union member that needs to be filled. But this correspondence is implicit and is not captured (left as an exercise to the reader), so fuzzer will be trying all 60 possible combinations of cmd x union member, while only 10 are valid.
So the transpose would be something along the following lines:
```
type vlan_args[CMD, DATA] {
	cmd		const[CMD, intptr]
	device1		array[int8, 24]
	data		vlan_args_u[DATA]
	vlan_qos	int8
}

type vlan_args_u[DATA] {
	data DATA
} [size[24]]
```

And then we can actually match cmd with data as e.g.:
```
ioctl$sock_SIOCSIFVLAN_ADD_VLAN_CMD(fd sock, cmd const[SIOCSIFVLAN], arg ptr[in, vlan_args[ADD_VLAN_CMD, string[devnames, 24]]])
```

Reply:
That makes a lot of sense, really didn't think that much while adding the description. Updated.
One question is that a good idea to set the ptr dir to out and save the fuzzer power if the ptr is not going to be used for some command(GET_VLAN_INGRESS_PRIORITY_CMD and GET_VLAN_EGRESS_PRIORITY_CMD)? As there is a copy_from_user here: https://elixir.bootlin.com/linux/latest/source/net/8021q/vlan.c#L547, just passing a null pointer might not work, and will an out pointer be assigned or can be risky set to 0 at the beginning? I haven't have enough time to run fuzz but will keep it running for a while and check if the coverage is as expected tomorow. At the meantime, I will just update this PR first, if there is any problem I will keep working on it.

And actually as the union are all int32 except an string element, some of the particular commands seems to be able to combine, however it seems strange to have only some of them group together, so I just put every one of them separately. Let me know if that doesn't look good.